File: Data Structures & Algorithms/combination-target-sum-ii/submission-4.py
from typing import List

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        ans = []
        def dfs(i: int, subset: List[int], total: int):
            if total == target:
                ans.append(subset.copy())
                return
            elif total > target:
                return
            
            if i == len(candidates):
                return
            
            subset.append(candidates[i])
            total += candidates[i]
            dfs(i + 1, subset, total)
            
            subset.pop()
            total -= candidates[i]
            curr = candidates[i]
            while i < len(candidates) and curr == candidates[i]:
                i += 1
            dfs(i, subset, total)
            
        dfs(0, [], 0)
        return ans

File: Data Structures & Algorithms/coin-change-ii/submission-4.cpp
class Solution
{
public:
    int change(int amount, vector<int>& coins) {
        sort(coins.begin(), coins.end());
        vector<vector<int>> dp(coins.size(), vector<int>(amount + 1, -1));
        return dfs(coins, 0, amount, dp);
    }

    int dfs(const vector<int> coins, int i, int a, vector<vector<int>>& dp) {
        if (a == 0) {
            return 1;
        }

        if (i >= coins.size()) {
            return 0;
        }

        if (dp[i][a] != -1) {
            return dp[i][a];
        }

        int ans = 0;
        if (a >= coins[i]) {
            ans = dfs(coins, i + 1, a, dp);
            ans += dfs(coins, i, a - coins[i], dp);
        }
        dp[i][a] = ans;
        return ans;
    }
};